demos/slack-alert-app/slackbot.py

```diff
@@ -0,0 +1,137 @@
+# limitations under the License.
+
+import time
+import logging
+import os
+import schedule
+import pxapi
+from slack.web.client import WebClient
+from slack.errors import SlackApiError
+
+
+# Pixie PxL script:
+# This script ouputs a table of the HTTP requests count and
+# HTTP error (>4xxx) count for each service in the `px-sock-shop` namespace.
+PXL_SCRIPT = """
+import px
+
+df = px.DataFrame(table='http_events', start_time='-5m')
+
+# Add column for HTTP response status errors.
+df.error = df.http_resp_status >= 400
+
+# Add columns for service, namespace info
+df.namespace = df.ctx['namespace']
+df.service = df.ctx['service']
+
+# Filter for px-sock-shop namespace only.
+df = df[df.namespace == 'px-sock-shop']
+
+# Group HTTP events by service, counting errors and total HTTP events.
+df = df.groupby(['service']).agg(
+    error_count=('error', px.sum),
+    total_requests=('http_resp_status', px.count)
+)
+
+px.display(df, "status")
+"""
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+# Get data from the Pixie API.
+def get_pixie_data(cluster_conn):
+
+    # Execute the PxL script.
+    script = cluster_conn.prepare_script(PXL_SCRIPT)
+    logging.debug("Pixie cluster executed script.")
+
+    service_stats_msg = ["*Recent 4xx+ Spikes in last 5 minutes:*"]
+
+    # Process table output rows to construct slack message.
+    for row in script.results("status"):
+        service_stats_msg.append(format_message(row["service"],
+                                                row["total_requests"],
+                                                row["error_count"]))
+
+    return "\n\n".join(service_stats_msg)
+
+
+# Format Pixie API table row data.
+def format_message(service, request_count, error_count):
+    return (f"`{service}` \t ---> {error_count} (>4xx)"
+            f" errors out of {request_count} requests.")
+
+
+# Send Slack message through the Slack client.
+def send_message(slack_client, channel, cluster_conn):
+
+    # Get data from the Pixie API.
+    msg = get_pixie_data(cluster_conn)
+
+    # Send a POST request through the Slack client.
+    try:
+        logging.info(f"Sending {msg!r} to {channel!r}")
+        slack_client.chat_postMessage(channel=channel, text=msg)
+
+    except SlackApiError as e:
+        logging.error('Request to Slack API Failed: {}.'.format(e.response.status_code))
+        logging.error(e.response)
+
+
+if __name__ == "__main__":
+
+    # Get Pixie API key.
+    if "PIXIE_API_KEY" not in os.environ:
+        logging.error("Missing `PIXIE_API_KEY` environment variable.")
+    PIXIE_API_KEY = os.environ['PIXIE_API_KEY']
+
+    # Create a Pixie client.
+    logging.debug("Authorizing Pixie client.")
+    px_client = pxapi.Client(token=PIXIE_API_KEY)
+
+    # Get Pixie cluster ID.
+    if "PIXIE_CLUSTER_ID" not in os.environ:
+        logging.error("Missing `PIXIE_CLUSTER_ID` environment variable.")
+    PIXIE_CLUSTER_ID = os.environ['PIXIE_CLUSTER_ID']
+
+    # Connect to cluster.
```

Comment:
I usually find comments that just refer to the code not that useful. For example:
    # Connect to cluster.

   connect_to_cluster(...)

It's better to have more descriptive comments than define one for each line. 


For example, 

\# The slackbot needs the following configs: cluser_id, slack_bot_token ...
\# We allow these to be specified using the environment variables...

Then the code to fetch all the tokens.